execution/nodeInterface/NodeInterface.go

```diff
@@ -53,90 +51,126 @@ var merkleTopic common.Hash
 var l2ToL1TxTopic common.Hash
 var l2ToL1TransactionTopic common.Hash
 
-var blockInGenesis = errors.New("")
-var blockAfterLatestBatch = errors.New("")
-
 func (n NodeInterface) NitroGenesisBlock(c ctx) (huge, error) {
 	block := n.backend.ChainConfig().ArbitrumChainParams.GenesisBlockNum
 	return arbmath.UintToBig(block), nil
 }
 
+// bool will be false but no error if behind genesis
+func (n NodeInterface) blockNumToMessageIndex(blockNum uint64) (arbutil.MessageIndex, bool, error) {
+	node, err := gethExecFromNodeInterfaceBackend(n.backend)
+	if err != nil {
+		return 0, false, err
+	}
+	blockchain, err := blockchainFromNodeInterfaceBackend(n.backend)
+	if err != nil {
+		return 0, false, err
+	}
+	if blockNum < blockchain.Config().ArbitrumChainParams.GenesisBlockNum {
+		return 0, true, nil
+	}
+	msgIndex, err := node.ExecEngine.BlockNumberToMessageIndex(blockNum)
+	if err != nil {
+		return 0, false, err
+	}
+	return msgIndex, true, nil
+}
+
+func (n NodeInterface) msgNumToInboxBatch(msgIndex arbutil.MessageIndex) (uint64, bool, error) {
+	node, err := gethExecFromNodeInterfaceBackend(n.backend)
+	if err != nil {
+		return 0, false, err
+	}
+	fetcher := node.ExecEngine.GetBatchFetcher()
+	if fetcher == nil {
+		return 0, false, errors.New("batch fetcher not set")
+	}
+	return fetcher.FindInboxBatchContainingMessage(msgIndex)
+}
+
 func (n NodeInterface) FindBatchContainingBlock(c ctx, evm mech, blockNum uint64) (uint64, error) {
-	node, err := arbNodeFromNodeInterfaceBackend(n.backend)
+	msgIndex, found, err := n.blockNumToMessageIndex(blockNum)
 	if err != nil {
 		return 0, err
 	}
-	return findBatchContainingBlock(node, node.TxStreamer.GenesisBlockNumber(), blockNum)
+	if !found {
+		return 0, fmt.Errorf("block %v is part of genesis", blockNum)
+	}
+	res, found, err := n.msgNumToInboxBatch(msgIndex)
+	if err == nil && !found {
+		return 0, errors.New("block not yet found on any block")
```

Comment:
should be something like "block not yet found in any batch". currently says "block" twice